miner/block_builder.go

```diff
@@ -194,9 +195,17 @@ func (t *BlockBuilder) createBlock(id types.LayerID, atxID types.AtxId, eligibil
 				log.Uint64("bottom", uint64(bottom)), log.Uint64("top", uint64(top)),
 				log.Uint64("hdist", uint64(t.hdist)), log.Err(err))
 		}
-		if votes == nil { // if no votes set to empty
-			t.Info("Votes is nil. Setting votes to an empty array")
-			votes = []types.BlockID{}
+		if votes == nil || len(votes) == 0 { // if no votes, vote as the whole layer
```

Comment:
change comment to "vote for orpahns as of this layer"

Reply:
Why this OR?

@antonlerner none of the blocks should be orphan at this time, since they are in the view of older blocks.

Reply:
@barakshani to my understanding, if hare results haven't arrived yet we should vote in the same way we would have proposed a set to the hare, which in this case should be the orphans that this nodes sees. even if CP hasn't didn't terminate it is not guaranteed that blocks were received that vote for these orphans

Reply:
using get layer is just a way of getting the orphans and none orphans of the specific layer, since now we might be in the next layer getOrphans wont work

Reply:
@antonlerner There are 2 options, Gavrad will choose from which to implement:
1. if no have result for layer j vote neutral, unless the current layer is j+hdist, which is the last option to vote explicitly on layer j.
2. Never vote neutral, and if you have no results then vote according to your local view.

Option 2 is what you were referring to, while I was actually talking about option 1.
In both cases, the right thing to do is to vote similarly to the proposed set as you said. For now, it is ok just to take the layer (so if we got late blocks, they will be voted for as well).

Also, as Almog said, if we go with option 2, then in layer j+2 these blocks are not orphans.

Reply:
Note that the logic here is not complete: we should ask about Hare results _per layer_.
I talked with Gavrad about it.

Reply:
Regarding the case of j+2, what does that mean exactly? that we are trying to create a late block? i.e a block with layerid 100 in layer 101?

Reply:
No, in layer 102 you also vote explicitly about layer 100 (if hdist >1), but if you have no hare results about layer 100 yet, then calling `get orphans` for layer 100 would (should) return an empty set.

Reply:
I don't think this is what this code does. this code creates a block in layer `id`, if there are no votes between `top` to `bottom` which is hdist, the block will vote for the blocks in layer `id` and not `bottom` as you are described in your example of 102 and trying to call getOrphans for 100
@gavraz correct me if I'm wrong

Reply:
yes, this is not what the code does:
1. Blocks from layer id should not vote about blocks from layer id.
2. The logic should be as I described.

Gavrad knows, we talked about it.

Reply:
Fixed

Reply:
what about this line here `l, e := t.orphans.GetLayer(id)` will it take only the last layer?